server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -325,6 +347,137 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
+     * If vendor plugin supports uploading multiple parts in parallel, <code>BlobContainer#writeBlobByStreams</code>
+     * will be used, else, the legacy {@link RemoteSegmentStoreDirectory#copyFrom(Directory, String, String, IOContext, boolean)}
+     * will be called.
+     *
+     * @param from          The directory for all files to be uploaded
+     * @param files         A list containing the names of all files to be uploaded
+     * @param context       IOContext to be used to open IndexInput to files during remote upload
+     * @param uploadTracker An {@link UploadTracker} for tracking file uploads
+     * @throws Exception When upload future creation fails or if {@link RemoteSegmentStoreDirectory#copyFrom(Directory, String, String, IOContext, boolean)}
+     *                   throws an exception
+     */
+    public boolean copyFilesFrom(Directory from, Collection<String> files, IOContext context, UploadTracker uploadTracker)
+        throws Exception {
+
+        List<CompletableFuture<Void>> resultFutures = new ArrayList<>();
+
+        boolean uploadOfAllFilesSuccessful = true;
+        for (String src : files) {
+            String remoteFilename = createRemoteFileName(src, false);
+            uploadTracker.beforeUpload(src);
+            if (remoteDataDirectory.getBlobContainer().isMultiStreamUploadSupported()) {
+                try {
+                    CompletableFuture<Void> resultFuture = createUploadFuture(from, src, remoteFilename, context);
+                    resultFuture.whenComplete((uploadResponse, throwable) -> {
+                        if (throwable != null) {
+                            uploadTracker.onFailure(src);
+                        } else {
+                            uploadTracker.onSuccess(src);
+                        }
+                    });
+                    resultFutures.add(resultFuture);
+                } catch (Exception e) {
+                    uploadTracker.onFailure(src);
+                    throw e;
+                }
+            } else {
+                boolean success = true;
+                try {
+                    copyFrom(from, src, src, context, false);
+                    uploadTracker.onSuccess(src);
+                } catch (IOException e) {
+                    success = false;
+                    uploadOfAllFilesSuccessful = false;
+                    logger.warn(() -> new ParameterizedMessage("Exception while uploading file {} to the remote segment store", src), e);
+                } finally {
+                    if (!success) {
+                        uploadTracker.onFailure(src);
+                    }
+                }
+            }
+        }
+
+        if (resultFutures.isEmpty() == false) {
+            CompletableFuture<Void> resultFuture = CompletableFuture.allOf(resultFutures.toArray(new CompletableFuture[0]));
+            try {
+                resultFuture.get();
```

Comment:
Which thread will block on the transfer?

Reply:
Main thread which must be translog sync thread I think. Need to confirm this.